[LSP] Add Razor options provider to Roslyn

allisonchou opened the pull request:
- Roslyn isn't respecting Razor's tabs/spaces settings when generating override and partial method completion TextEdits in the resolve handler.
- This PR adds a document options provider that allows Razor to communicate its options to Roslyn.
- The Razor side of this also needs to be merged for this bug to be resolved: https://github.com/dotnet/aspnetcore-tooling/pull/3719

Fixes (once Razor side is also merged):
https://github.com/dotnet/aspnetcore/issues/32555

Youssef1313 commented on src/Features/Core/Portable/Completion/CompletionProvider.cs:
```diff
@@ -69,11 +69,20 @@ public virtual Task<CompletionDescription> GetDescriptionAsync(Document document
+        /// <param name="item">The item to be committed.</param>
+        /// <param name="commitKey">The optional key character that caused the commit.</param>
+        /// <param name="cancellationToken"></param>
+        internal virtual Task<CompletionChange> GetChangeAsync(Document document, OptionSet optionSet, CompletionItem item, char? commitKey, CancellationToken cancellationToken)
+            => GetChangeAsync(document, item, commitKey, cancellationToken);
```
@333fred FYI about the new internal `GetChangeAsync` overload.

333fred replied:
Adding a new overload concerns me. We only just removed the last one with https://github.com/dotnet/roslyn/pull/53261, and this has the same opportunity for us to break the public API version again.

allisonchou replied:
I removed the overload and took a different approach, so I don't think this should be a concern anymore

davidwengier commented on src/Features/LanguageServer/Protocol/Handler/Completion/CompletionResolveHandler.cs:
```diff
@@ -185,5 +188,40 @@ private static bool MatchesLSPCompletionItem(LSP.CompletionItem lspCompletionIte
+                return documentOptions;
+            }
+
+            if (options.TryGetDocumentOption(new OptionKey(FormattingOptions.UseTabs, document.Project.Language), out var useTabs) &&
+                useTabs != null &&
+                useTabs.GetType() == typeof(bool))
+            {
+                documentOptions = documentOptions.WithChangedOption(FormattingOptions.UseTabs, (bool)useTabs);
+            }
+
+            if (options.TryGetDocumentOption(new OptionKey(FormattingOptions.TabSize, document.Project.Language), out var tabSize) &&
+                tabSize != null &&
+                tabSize.GetType() == typeof(int))
+            {
+                documentOptions = documentOptions.WithChangedOption(FormattingOptions.TabSize, (int)tabSize);
+                documentOptions = documentOptions.WithChangedOption(FormattingOptions.IndentationSize, (int)tabSize);
+            }
```
Seems odd that these three specific options are hard-coded here. Should this be moved to the Razor-specific code, as that is the thing that knows that Razor cares about these three options.

dibarbet commented on src/Features/LanguageServer/Protocol/Handler/Completion/CompletionResolveHandler.cs:
```diff
@@ -185,5 +188,40 @@ private static bool MatchesLSPCompletionItem(LSP.CompletionItem lspCompletionIte
         }
+
+        // Certain language servers such as Razor may want TextEdits formatted using their own options instead of C#/VB options.
+        private static async Task<OptionSet> GetCompletionFormattingOptionsAsync(Document document, CancellationToken cancellationToken)
```
A document options provider for razor documents seems like a sound approach to me, but if we're going that route, I think we should bake it into where the options are used (e.g. the formatter).  
That way any LSP feature that cares about razor options just does the right thing automatically.  With the current implementation, we have to repeat this code on every LSP feature and implementers have to be aware of the razor case otherwise it is by default broken.

dibarbet reviewed (commented):
Have one main piece of feedback before I look at the rest

allisonchou commented on src/Features/Core/Portable/Completion/CompletionProvider.cs:
```diff
@@ -69,11 +69,8 @@ public virtual Task<CompletionDescription> GetDescriptionAsync(Document document
-            return Task.FromResult(CompletionChange.Create(new TextChange(item.Span, item.DisplayText)));
-        }
+        public virtual Task<CompletionChange> GetChangeAsync(Document document, CompletionItem item, char? commitKey, CancellationToken cancellationToken)
+            => Task.FromResult(CompletionChange.Create(new TextChange(item.Span, item.DisplayText)));
```
I just changed this to expression body to be more consistent with the style of the other methods in this class

allisonchou commented:
I totally overhauled this PR based on your super helpful feedback @davidwengier and @dibarbet. The check for Razor options is now performed in the formatter, and the specific tabs/spaces options logic is now done on the Razor side (via https://github.com/dotnet/aspnetcore-tooling/pull/3719). If either of you could take another look whenever you have time, that would be appreciated. Thanks 😀

davidwengier reviewed (approved):
So much cleaner 😁

dibarbet reviewed (approved):
🎉 totally agree with David, I much prefer this iteration

NTaylorMullen commented on src/Workspaces/Core/Portable/Formatting/Formatter.cs:
```diff
@@ -146,6 +146,14 @@ internal static async Task<Document> FormatAsync(Document document, SyntaxAnnota
+            var documentOptionSetProvider = document.Services.GetService<IDocumentOptionSetProvider>();
+            if (documentOptionSetProvider is not null)
+            {
+                options = await documentOptionSetProvider.GetOptionsForDocumentAsync(document, cancellationToken).ConfigureAwait(false);
```
Hmmm, one interesting implication is that if you pass in options to this method they'll be ignored because the document options will win. With that in mind the options on line 157 account for this; in an ideal world all call sites to this method wouldn't actually pass options in "within" Roslyn and would then fall back to the appropriate document options services. All that can be done in the future though, just my 2cents

jasonmalinowski replied:
Yeah this seems really weird here. Couldn't the LSP layer or something deal with coupling this at that layer versus something else?

allisonchou replied:
Currently Razor is the only implementer of `IDocumentSetProvider` so in practice I don't think there's currently an issue, however I totally see both of your points about the check being weird here. I'm exploring alternate options, will hold off on merging in the meantime.

jasonmalinowski commented on src/Workspaces/Core/Portable/Formatting/Formatter.cs:
```diff
@@ -146,6 +146,14 @@ internal static async Task<Document> FormatAsync(Document document, SyntaxAnnota
 
             var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            // Languages such as Razor can specify their own options instead of using the document's options.
+            var documentOptionSetProvider = document.Services.GetService<IDocumentOptionSetProvider>();
+            if (documentOptionSetProvider is not null)
```
So why is this having to be checked _here_ rather than in the core document options logic we have, like through https://github.com/dotnet/roslyn/blob/main/src/Workspaces/Core/Portable/Options/IDocumentOptionsProvider.cs?

jasonmalinowski reviewed (commented):
This approach feels *really* funky and it feels like we still don't have this wired up at the right layer....?

allisonchou commented:
@jasonmalinowski @NTaylorMullen Based on your feedback, I refactored out the Razor options logic from the formatter [over to the indentation service](https://github.com/dotnet/roslyn/pull/53879/files#diff-e3a49cfd1307372020dfcc2e94cfcba14a6cf75869149f13da33f1301bba0e3aR43). This required once again overhauling a large portion of the PR, namely due to having to [change the return type](https://github.com/dotnet/roslyn/pull/53879/files#diff-22b55d82504bb8a6d152b39ca765d9ae7d2e7889d54ecf176fb7b9a7e2d8ee8cR12) of the added Razor external access method. If either of you (or anyone else reading this) has any concerns with the updated approach, please let me know!

NTaylorMullen commented on src/Tools/ExternalAccess/Razor/RazorDocumentOptionsServiceWrapper.cs:
```diff
@@ -0,0 +1,56 @@
+
+            public RazorDocumentOptions(IRazorDocumentOptions razorOptions)
+            {
+                _tryGetDocMethod = razorOptions.TryGetDocumentOption;
```
Instead of holding onto the method why not hold onto the options and do `_razorOptions.TryGetDocumentOptions` below?

allisonchou replied:
ah good point! Updated, thanks 👍

jasonmalinowski commented on src/EditorFeatures/TestUtilities/LanguageServer/AbstractLanguageServerProtocolTests.cs:
```diff
@@ -68,10 +69,40 @@ public void Register(Workspace workspace)
             TService IDocumentServiceProvider.GetService<TService>()
-                => (TService)(object)new TestSpanMapper();
+            {
+                if (typeof(IDocumentOptionsService).IsAssignableFrom(typeof(TService)))
```
Can't we just do == here? I assume we're not asking for derived types?

jasonmalinowski commented on src/EditorFeatures/TestUtilities/LanguageServer/AbstractLanguageServerProtocolTests.cs:
```diff
@@ -68,10 +69,40 @@ public void Register(Workspace workspace)
+
+            private class TestDocumentOptions : IDocumentOptions
+            {
+                public bool TryGetDocumentOption(OptionKey option, out object? value)
+                {
+                    value = null;
+                    return false;
+                }
```
If this is just a no-op do we really need to add it?

jasonmalinowski commented on src/EditorFeatures/Core/Implementation/Formatting/InferredIndentationDocumentOptionsProviderFactory.cs:
```diff
@@ -38,8 +38,21 @@ private class DocumentOptionsProvider : IDocumentOptionsProvider
-                => Task.FromResult<IDocumentOptions?>(new DocumentOptions(document.Project.Solution.Workspace, document.Id, _indentationManagerService));
+            public async Task<IDocumentOptions?> GetOptionsForDocumentAsync(Document document, CancellationToken cancellationToken)
+            {
+                // Languages such as Razor can specify their own formatting options.
```
So it's a bit odd that we are having to modify the InferredIndentationDocumentOptionsProviderFactory here when we could have a new MEF export. The Razor ExternalAccess could directly export it's own IDocumentOptionsProviderFactory that imports the Razor service directly. This would mean you can keep almost the entire change in the Razor layer and not have to introduce the extra layer in the middle: just rename IDocumentOptionsService to IRazorDocumentOptionsService and you can get away with bit less wrapping.

jasonmalinowski replied:
(it's also otherwise odd we'd be having this at an EditorFeatures layer, since what happens for VS Code?)

allisonchou commented on src/Features/Core/Portable/Shared/Extensions/DocumentExtensions.cs:
```diff
@@ -179,5 +179,13 @@ public static ImmutableArray<AbstractFormattingRule> GetFormattingRules(this Doc
             return rules.AddRange(Formatter.GetDefaultFormattingRules(document));
         }
+
+        public static bool IsRazorDocument(this Document document)
```
Method was moved

NTaylorMullen replied:
You moved it here for ExternalAccess.Razor right? If you all are "ok" with this living here then I don't have an issue but I also wouldn't be opposed to duplicating this in the ExternalAccess.Razor layer

allisonchou replied:
Created a new ExternalAccess class for Razor extensions - https://github.com/dotnet/roslyn/blob/b9eb126fc9c7539f6e7cd3ee31dd1da318ac2f1c/src/Tools/ExternalAccess/Razor/Extensions.cs

Hope this is a little cleaner!

allisonchou commented:
@jasonmalinowski I moved to a MEF-based approach in response to your feedback. One thing I was unsure about was whether there's a way to exclusively activate the Razor options factory in Razor scenarios, as right now it activates with every type of project/solution. To mitigate this, I added a `IsRazorDocument` check to the Razor options provider, but I'm not sure if there's a better approach.

ryanbrandenburg commented on src/Tools/ExternalAccess/Razor/IRazorDocumentOptionsService.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.Razor
+{
+    using System.Threading;
```
Nit: Seems like the style of the rest of this is using outside the namespace.

allisonchou replied:
Thanks for the catch! You're right, I moved them up.

ryanbrandenburg reviewed (approved):
One nit-pick, otherwise looks OK, though I lack context.

allisonchou commented on src/Tools/ExternalAccess/Razor/RazorDocumentOptionsProviderFactory.cs:
```diff
@@ -0,0 +1,76 @@
+namespace Microsoft.CodeAnalysis.ExternalAccess.Razor
+{
+    [Shared]
+    [ExportMetadata("Extensions", new string[] { "cshtml", "razor", })]
```
Great point, I checked with Taylor just now and it seems the line is actually unnecessary so I removed it - seems like it's required for Razor's dynamic file info provider but not here, it just tells Roslyn to ask Razor for dynamic file info when a .cshtml or .razor file is added.

allisonchou commented:
Thanks everyone for the reviews! I think I've gotten to all the feedback, if there's anything else please let me know. If everything looks good, I plan to merge this by EOD so we can start getting the Razor side PR going as well.
